### DISCUSSION TITLE
flightmode.stabModeRoll

### DISCUSSION THREAD

**Discussion by fangjir** [bitcraze/discussions#1307] [Q&A]
i know we can change  the value of flightmode.stabModeRoll for 0 for latest firmware, does this means when we use send_setpoint(roll,pitch,yaw,thrust),here the roll is the rate ? before i use the send_setpoint for angle without changing stabModeRoll.
i want to know this because i want to design the attitude controller.
![屏幕截图 2024-04-26 145249](https://github.com/bitcraze/discussions/assets/71303114/18f188b6-ef43-46f1-9020-630083c2982e)

**Comment by fangjir** [comment]
and I also want to know what the rotation order of Euler angles looks like in the rotation matrix in crazyflie?

**Reply by gemenerik** [reply]
Since this is an unrelated question, please ask it in a new discussion.

**Comment by gemenerik** [comment]
Yes. In `commander.py` in cflib:
```python
def send_setpoint(self, roll, pitch, yawrate, thrust):
    """
    Send a new control setpoint for roll/pitch/yaw_Rate/thrust to the copter.

    The meaning of these values is depended on the mode of the RPYT commander in the firmware
    Default settings are Roll, pitch, yawrate and thrust
    ...
    """
```

and in `crtp_commander_rpyt.c`:
```c
typedef enum
{
  RATE    = 0,
  ANGLE   = 1,
} RPYType;

...

static RPYType stabilizationModeRoll  = ANGLE; // Current stabilization type of roll (rate or angle)
static RPYType stabilizationModePitch = ANGLE; // Current stabilization type of pitch (rate or angle)
static RPYType stabilizationModeYaw   = RATE;  // Current stabilization type of yaw (rate or angle)
```

These can be set with the parameter framework. Setting one of these to 0 thus indeed makes the commander expect rates instead of angles.
